src/lib.cairo

```diff
@@ -67,6 +107,182 @@ mod Blobstream {
+            // Check that current validators and signatures are well-formed.
+            assert(
+                _current_validator_set.len() == _sigs.len(), Errors::MALFORMED_CURRENT_VALIDATOR_SET
+            );
+
+            // Check that the supplied current validator set matches the saved checkpoint.
+            let current_validator_set_hash: u256 = compute_validator_set_hash(
+                _current_validator_set
+            );
+            let domain_separate_validator_set_hash_val = domain_separate_validator_set_hash(
+                _old_nonce, current_power_threshold, current_validator_set_hash
+            );
+            assert(
+                domain_separate_validator_set_hash_val == last_validator_set_checkpoint,
+                Errors::SUPPLIED_VALIDATOR_SET_INVALID
+            );
+
+            // Check that enough current validators have signed off on the new validator set.
+            let new_checkpoint: u256 = domain_separate_validator_set_hash(
+                _new_nonce, _new_power_threshold, _new_validator_set_hash
+            );
+            check_validator_signatures(
+                @self, _current_validator_set, _sigs, new_checkpoint, current_power_threshold
+            );
+
+            self.state_last_validator_checkpoint.write(new_checkpoint);
+            self.state_power_threshold.write(_new_power_threshold);
+            self.state_event_nonce.write(_new_nonce);
+        // TODO Add event emission: ISSUE:  #25
+        }
+
+        fn submit_data_root_tuple_root(
+            ref self: ContractState,
+            _new_nonce: felt252,
+            _validator_set_nonce: felt252,
+            _data_root_tuple_root: u256,
+            _current_validator_set: Span<Validator>,
+            _sigs: Span<Signature>
+        ) {
+            let current_nonce: felt252 = self.state_event_nonce.read();
+            let current_power_threshold: felt252 = self.state_power_threshold.read();
+            let last_validator_set_checkpoint: u256 = self.state_last_validator_checkpoint.read();
+
+            // Check that the new nonce is one more than the current one.
+            assert(_new_nonce == current_nonce + 1, Errors::INVALID_DATA_ROOT_TUPLE_ROOT_NONCE);
+            // Check that current validators and signatures are well-formed.
+            assert(
+                _current_validator_set.len() == _sigs.len(), Errors::MALFORMED_CURRENT_VALIDATOR_SET
+            );
+
+            // Check that the supplied current validator set matches the saved checkpoint.
+            let current_validator_set_hash: u256 = compute_validator_set_hash(
+                _current_validator_set
+            );
+
+            // Check that the supplied current validator set matches the saved checkpoint.
+            let current_validator_set_hash: u256 = compute_validator_set_hash(
+                _current_validator_set
+            ); // TODO: compute_validator_set_hash
```

Comment:
If this currently has an issue please comment the # if not please create and issue and comment